Flesh out custom layout doc; fix typos in #650 (#651)
Julow/Unexpected-Keyboard: doc/Custom-layouts.md

# Custom layouts

> Unexpected Keyboard includes key layouts for many natural languages. Unexpected starts out with a key layout likely to suit your device's language. You can select a different layout by calling up the Settings page (swipe the gear icon) and, at the top of the page, either tapping an existing layout or tapping _Add an alternate layout_. This displays a menu of available layouts.

Author: Should be addressed by my revision. I explain the need for such an intro in my follow-on post.

If no built-in layout is what you want, you can define your own layout by choosing _Custom layout_ at the bottom of this menu. Unexpected Keyboard now displays code in the XML format. You make changes by replacing this with different code and tapping OK.

Start out in one of these ways:

> * Copypaste the displayed code into a text file (named something like `MyChanges.xml`) to maintain it independently of Unexpected Keyboard.

Author: Should be resolved by my revision.

* Make a copy of one of the built-in layouts found in [`/srcs/layouts`](https://github.com/Julow/Unexpected-Keyboard/tree/master/srcs/layouts).

> * Use the [web-based editor](https://unexpected-keyboard-layout-editor.lixquid.com/). You can use touch gestures to define keys and swipes and move keys to desired positions, and it will write the XML code for you.

Author: Should be resolved by my revision (rewording). I mistakenly assumed the web page let you drag keys into position.

The `<keyboard>`...`</keyboard>` pair follows the declaration tag and encloses the whole keyboard. The following properties may be used (The first two appear in the example above):
* `name`: The name of the keyboard. The name you specify will appear in the Settings menu. If not present, the layout will just appear as “Custom layout”.
* `script`: The (main) writing system that the keyboard supports. The possible values are `arabic`, `armenian`, `bengali`, `cyrillic`, `devanagari`, `gujarati`, `hangul`, `hebrew`, `latin`, `persian`, `shavian`, and `urdu`. Defaults to `latin`.
* `numpad_script`: The script to use for the numpad. This is useful for scripts where a different, non-ASCII set of numerals is used, like Devanagari and Arabic. Defaults to the same as `script`.
* `bottom_row`: Whether or not to show the common bottom row. It accepts `true` or `false`, and defaults to `true`. If your custom layout defines a bottom row, you should specify `bottom_row=false` to disable the built-in bottom row.

## Row
The `<row>`...`</row>` pair encloses one row on the keyboard. It has only one optional property:
* `height`: The height of the row: a positive floating-point value.

> A row's standard height is 1.0. The `height` property makes the row taller or shorter than the others. For example, if you define a 5-row keyboard but the top row has `height="0.7"`, then the keyboard's total height is 4.7 units.

Reviewer: ```suggestion
A row's default height is 1.0. The `height` property makes the row taller or shorter than the others. For example, if you define a 5-row keyboard but the top row has `height="0.7"`, then the keyboard's total height is 4.7 units.
```

Author: Agreed, a more conventional term.

> You set the physical height of the complete keyboard in Settings as a percentage of the total height of the screen, which can be different between portrait and landscape. Unexpected Keyboard scales the "units" so all the rows fit in that alloted space. Thus, defining a shorter row gives space to the other rows.

Reviewer: That's not true. The percentage set in the setting is only true for a 4 row layout where every rows have a height of `1.0`. Adding or removing a row or changing a row's height will change the keyboard apparent size and will not scale the keys.

The interaction between unusual layouts and this option has never be thought of so I suggest dropping this paragraph.

## Key
The `<key />` tag defines a key on the keyboard. Its position in the sequence of keys inside `<row>`...`</row>` indicates its position in the row from left to right. It must have the following property:

> * `key0`: What the key does when tapped. Unexpected Keyboard provides a legend in the middle of the key's area.

Reviewer: This property is not mandatory! It's reasonable to have a key without a center symbol, for example a directional pad.

Author: Should be resolved by my revision.

The key tag may have additional properties as set out below.

### Swipes
The following optional properties define the effects of swipes:
* `n`, `ne`, `e`, `se`, `s`, `sw`, `w`, `nw`: What the key should do when it is swiped in the direction of that compass point. ("North" means upward and "East" is to the right.)

 nw | n    |  ne
:-: | :--: | :-:
 w  | key0 |  e
 sw | s    |  se

* `key1` through `key8` is an older way to achieve the same effects. The directions are ordered as follows:

key1 | key7 | key2
:-: | :--: | :-:
key5 | key0 | key6
key3 | key8 | key4

You can define a swipe only once with either compass-point or numeric notation. Unexpected Keyboard automatically puts a small legend in that direction from the center of the key.

> * `slider`: If `slider="true"`, and the key also has `w` and `e` properties, then the key tracks horizontal finger motion precisely and sends the `w` and `e` keystrokes repeatedly. In built-in layouts, this makes the space bar send left and right characters as the user slides on the space bar.

Reviewer: The internal mechanism is more complicated than that so let's not reveal unnecessary details.
```suggestion
* `slider`: If `slider="true"`, and the key also has `w` and `e` properties, then the key tracks horizontal finger motion precisely and sends the `w` and `e` keystrokes repeatedly. In built-in layouts, this is used to implement the space bar slider.
```

Author: Yes, fewer details on the flow of characters is better.

* `anticircle`: The key value to send when doing an anti-clockwise gesture on the key. (The clockwise gesture applies a Shift modifier and the round-trip gesture applies a Fn modifier. These results can be redefined using a modmap, as explained below.)

### Layout
A key may have the following properties to control the row's layout:
* `width`: The width of the key, a positive floating-point value. It defaults to 1.0
* `shift`: How much empty space to add to the left of this key, a non-negative floating-point value. It defaults to 0.0

Normally, a key's width is 1.0 unit. Unexpected Keyboard occupies the full width of the screen, and the row defining the highest number of units (in widths plus shifts) is as wide as the screen. A row whose width is a smaller number of units has empty space on the right.

### Extra legend
* `indication`: An optional extra legend to show under the main label. For example, `<key key0="2" indication="ABC" />` displays ABC at the bottom of the 2 key, as on a pinpad or some telephones. If the key also defines a downward swipe with `s` or `key8`, the legends overlap.

## Possible key values

> The properties that define the effects of tapping or swiping a key may be one of the built-in strings documented on [this page](Possible-key-values). For example, `se="cut"` says that a southeast swipe produces the "cut" character (Ctrl-C).

Reviewer: I'd rather not use "character" for talking of anything else than a piece of text containing only one visual character. Internally, the "cut" key is an "editing key" but I'd prefer not to over explain these internal details. "Key" is fine.
```suggestion
The properties that define the effects of tapping or swiping a key may be one of the built-in strings documented on [this page](Possible-key-values). For example, `se="cut"` says that a southeast swipe produces the "cut" key (Ctrl-C).
```

Author: Okay. Ctrl-C is a control "character" (or "code") but I agree, reader wants to deal with keys, not bytes.

Some of those strings begin with `loc `. These are place-holders; the tap or swipe does nothing unless enabled through the "Add keys to keyboard" option in the Settings menu, or implicitly enabled by the language the device is set to use. For example, `ne="loc accent_aigu"` says that a northeast swipe produces the acute accent combinatorial key—if enabled.

If the string defining a tap or a swipe is anything other than one of the built-in strings, the defined string is output *verbatim.* This is what most of the taps and swipes on a typical keyboard do. So `key0="a"` simply outputs the letter a.

## Modmap
The `<modmap>`...`</modmap>` pair encloses custom mappings for modifier keys.

A modmap can contain the following tags, each of which must have an `a` and a `b` property:

> * `<shift a="`...`" b="`...`" />` —This says that, if the Shift modifier is on (or the user made an anti-clockwise gesture on a key), and if the key would normally generate the "a" sequence, it must instead generate the  "b" sequence.

Reviewer: "Sequence" here is unclear, it can only be one key.
```suggestion
* `<shift a="`...`" b="`...`" />` —This says that, if the Shift modifier is on (or the user made a clockwise or round-trip gesture on a key), and if the key would normally send the key specified by "a", it must instead send the key specified by "b".
```
Same below for `fn`.

> * `<fn a="`...`" b="`...`" />` —This says that, if the Fn modifier is on (or the user made a round-trip gesture on a key), and if the key would normally generate the "a" sequence, it must instead generate the  "b" sequence.

Reviewer: The gestures are affected by both shift (preferably) and fn (as a fallback if shift did nothing).

# Custom layouts

> Unexpected Keyboard includes key layouts for several natural languages. Unexpected starts out with a key layout likely to suit your device's language; if there is none, Unexpected starts with a traditional QWERTY layout. You can select a different layout by calling up the Settings page (swipe the gear icon) and, at the top of the page, either tapping an existing layout or tapping _Add an alternate layout_. This displays a menu of available layouts.

Reviewer: I really don't think this paragraph is useful and I'd prefer if it was removed.

You can define your own layout by choosing _Custom layout_ at the bottom of this menu. Unexpected Keyboard now displays code in the XML format. You make changes by replacing this with different code and tapping OK.

> We recommend you keep your work in a file outside Unexpected Keyboard (named something like `MyChanges.xml`). If you installed a new version of Unexpected from a different website (with a different Build ID), then your Settings, including work you did solely by editing the XML inside Unexpected, would be lost.

Reviewer: ```suggestion
We recommend you keep your work in a file outside Unexpected Keyboard (named something like `MyChanges.xml`). If you installed a new version of Unexpected from a different website (with a different signature), then the work you did solely by editing the XML inside Unexpected, would be lost.
```
